### PR TITLE
Staging Sites: Add Staging Sites UI in Hosting Configuration

### PR DESCRIPTION
**Pull request by wojtekn**
Related to 1696-gh-Automattic/dotcom-forge

## Proposed Changes

In this PR, I propose to add an initial UI that allows adding a staging site.

## Testing Instructions

1. Open the Atomic site
2. Navigate to Hosting Configuration
3. Confirm that the "Staging site" card is displayed
4. Click "Create staging site" to add a new staging site
5. Confirm that when the site is created, the UI changes and allows navigating to staging site

Screenshot 2023-02-23 at 14 35 33

Screenshot 2023-02-23 at 14 34 41

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/my-sites/hosting/staging-site-card/use-check-staging-site-status.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+	useEffect( () => {
+		if ( ! siteId || transferStatus === transferStates.COMPLETE ) {
+			return;
+		}
+
+		let transferInterval: NodeJS.Timeout;
+
+		if ( ! isFetchingTransferStatus ) {
+			transferInterval = setInterval(
```
Comment (wojtekn) [inline client/my-sites/hosting/staging-site-card/use-check-staging-site-status.ts:26]: Initially, I tried to do it using `waitFor` like it's done in another component:

https://github.com/Automattic/wp-calypso/blob/f8823ab90e9a7b6dfa16c293abc32a8e35ca58aa/client/my-sites/checkout/checkout-thank-you/marketplace/marketplace-thank-you.tsx#L118

Unfortunately, it was sending tens of requests to the server. Do you have any thoughts on what could cause it? Does `waitFor` also need a way to cancel the effect?

**Review by wojtekn — COMMENTED** [review]
@Automattic/yolo please review the PR.

I would also appreciate looking at rendering, as I think that component still renders too often when I unfocus/focus the browser. I suspect I'm still missing something there. Or maybe I'm trying to do premature optimization?

File: client/my-sites/hosting/staging-site-card/use-check-staging-site-status.ts (hunk 1 above)
Comment (sejas) [reply client/my-sites/hosting/staging-site-card/use-check-staging-site-status.ts:26]: Screenshot 2023-02-23 at 20 58 13

During my testint, the `setInterval` worked great. It waited 5-6 seconds between each request, and it only made 3 requests.

**Review by sejas — APPROVED** [review]
It works great, I was able to create my staging site without errors. Love it! Great work!  🥳
I had to sandbox the API D102075-code

Some suggestions for future follow-ups/discussions related to the UI:
- We could set a minimum height to the `staging-site-card` so the card linking won't break the scrolling.
- Both buttons `Create staging site` and `Manage staging site` are too similar, which can confuse the user. We could change the text, or use the secondary button style for manage the site, or add some icons.
- Currently there is no easy way to "remove/refresh" the staging site.

Screenshot 2023-02-23 at 21 17 45

File: client/my-sites/hosting/staging-site-card/index.js [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+				<Button primary href={ `/home/${ stagingSiteSlug }` } disabled={ disabled }>
+					<span>{ translate( 'Manage staging site' ) }</span>
+				</Button>
+			</div>
+		);
+	};
+
+	return (
+		<Card className="staging-site-card">
+			<MaterialIcon icon="build" size={ 24 } />
```
Comment (sejas) [inline client/my-sites/hosting/staging-site-card/index.js]: This icon is conflicting with `Web server settings`

Screenshot 2023-02-23 at 21 27 26

File: config/stage.json [hunk 3]
```diff
@@ -155,6 +155,7 @@
 		"woop": false,
 		"wordpress-action-search": false,
 		"wpcom-user-bootstrap": true,
+		"yolo/staging-sites-i1": true,
```
Comment (sejas) [inline config/stage.json]: @wojtekn , Maybe we should wait a bit before adding the feature to stage?
My concerns are: HE recommending this feature, and people creating sites without any plan.

File: client/my-sites/hosting/staging-site-card/index.js (hunk 2 above)
Comment (wojtekn) [reply client/my-sites/hosting/staging-site-card/index.js]: Good catch! I tried to re-use an icon we already have in the codebase and missed that we use this one on this page. Would you have any thoughts on a good fit?

Screenshot 2023-02-24 at 10 33 52 Screenshot 2023-02-24 at 10 33 39 Screenshot 2023-02-24 at 10 33 33 Screenshot 2023-02-24 at 10 33 25

**Comment by wojtekn** [comment]
> Some suggestions for future follow-ups/discussions related to the UI:

> - We could set a minimum height to the staging-site-card so the card linking won't break the scrolling.
> - Both buttons Create staging site and Manage staging site are too similar, which can confuse the user. We could change the text, or use the secondary button style for manage the site, or add some icons.
> - Currently there is no easy way to "remove/refresh" the staging site.

Thanks @sejas , those are good suggestions. I've already added a ticket for the last one: 1757-gh-Automattic/dotcom-forge  and added another one for card height 1840-gh-Automattic/dotcom-forge

**Comment by wojtekn** [comment]
> Both buttons Create staging site and Manage staging site are too similar, which can confuse the user. We could change the text, or use the secondary button style for manage the site, or add some icons.

@javierarce any thoughts on this? Note that the "Manage staging site" button will be accompanied by "Delete staging site" button

File: client/my-sites/hosting/staging-site-card/index.js (hunk 2 above)
Comment (wojtekn) [reply client/my-sites/hosting/staging-site-card/index.js]: CC @javierarce

File: client/my-sites/hosting/main.js [hunk 4]
```diff
@@ -181,6 +184,7 @@ class Hosting extends Component {
 					<WrapperComponent>
 						<Layout className="hosting__layout">
 							<Column type="main" className="hosting__main-layout-col">
+								{ isStagingSiteEnabled && <StagingSiteCard disabled={ isDisabled } /> }
```
Comment (danielbachhuber) [inline client/my-sites/hosting/main.js]: I think we should put this under `<PhpMyAdminCard>` to start.

File: client/my-sites/hosting/staging-site-card/index.js [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+	const transferStatus = useCheckStagingSiteStatus( stagingSite.id );
+	const isStagingSiteTransferComplete = transferStatus === transferStates.COMPLETE;
+
+	const { addStagingSite, isLoading: addingStagingSite } = useAddStagingSiteMutation( siteId, {
+		onMutate: () => {
+			dispatch( removeNotice( stagingSiteAddFailureNoticeId ) );
+		},
+		onSuccess: () => {
+			dispatch( recordTracksEvent( 'calypso_hosting_configuration_staging_site_add_success' ) );
+			dispatch( successNotice( __( 'Staging site was added.' ), noticeOptions ) );
```
Comment (danielbachhuber) [inline client/my-sites/hosting/staging-site-card/index.js]:
```suggestion
			dispatch( successNotice( __( 'Staging site added.' ), noticeOptions ) );
```

This notice disappeared really quickly for me (much quicker than 3 seconds). Maybe we should make it always present?

File: config/stage.json [hunk 6]
```diff
@@ -155,6 +155,7 @@
 		"woop": false,
 		"wordpress-action-search": false,
 		"wpcom-user-bootstrap": true,
+		"yolo/staging-sites-i1": false,
```
Comment (danielbachhuber) [inline config/stage.json]: Let's simply remove the entry entirely.

**Review by danielbachhuber — COMMENTED** [review]
This is pretty great, @wojtekn. Nice work!

File: client/my-sites/hosting/staging-site-card/index.js (hunk 5 above)
Comment (wojtekn) [reply client/my-sites/hosting/staging-site-card/index.js]: Actually, now I see we don't need this notice. We display it when the simple site is added, but we still need to wait and poll for the transfer status. What if we remove it completely and solve it under 1767-gh-Automattic/dotcom-forge ?

Comment (danielbachhuber) [reply client/my-sites/hosting/staging-site-card/index.js]: @wojtekn I think a notice would be nice. Can we have it displayed after the staging site is fully created?

File: client/my-sites/hosting/staging-site-card/index.js (hunk 2 above)
Comment (sejas) [reply client/my-sites/hosting/staging-site-card/index.js]: We could also use Gutenberg icons: https://wordpress.github.io/gutenberg/?path=/story/icons-icon--library

Some suggestions:
Screenshot 2023-02-27 at 11 06 18

Screenshot 2023-02-27 at 11 07 04
Screenshot 2023-02-27 at 11 09 04

Screenshot 2023-02-27 at 11 10 58


Screenshot 2023-02-27 at 11 08 32


We also have the Gridicons: https://automattic.github.io/gridicons/

File: client/my-sites/hosting/staging-site-card/index.js (hunk 5 above)
Comment (danielbachhuber) [reply client/my-sites/hosting/staging-site-card/index.js]: Moved with 6ef8d1ee1e6e0027a007e18a3e298a9cfd1a41c3

File: client/my-sites/hosting/staging-site-card/index.js (hunk 2 above)
Comment (danielbachhuber) [reply client/my-sites/hosting/staging-site-card/index.js]: Going with the Share icon for now:

image

I created a follow up issue to explore further https://github.com/Automattic/dotcom-forge/issues/1856

**Review by danielbachhuber — APPROVED** [review]
Nice work on this, @wojtekn !

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7833935

Thank you @wojtekn for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7833935

Thank you @wojtekn for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7833935).

**Pull request merged**